nomad/deploymentwatcher/deployments_watcher.go

```diff
@@ -0,0 +1,414 @@
+// returned.
+func (w *Watcher) forceAdd(dID string) (*deploymentWatcher, error) {
+	// Build the request
+	args := &structs.DeploymentSpecificRequest{DeploymentID: dID}
+	var resp structs.SingleDeploymentResponse
+	if err := w.stateWatchers.GetDeployment(args, &resp); err != nil {
+		return nil, err
+	}
+
+	if resp.Deployment == nil {
+		return nil, fmt.Errorf("unknown deployment %q", dID)
+	}
+
+	return w.addLocked(resp.Deployment)
+}
+
+// getWatcher returns the deployment watcher for the given deployment ID.
+func (w *Watcher) getWatcher(dID string) (*deploymentWatcher, error) {
+	w.l.Lock()
+	defer w.l.Unlock()
+
+	// Not enabled so no-op
+	if !w.enabled {
+		return nil, notEnabled
+	}
+
+	watcher, ok := w.watchers[dID]
+	if ok {
+		return watcher, nil
+	}
+
+	return w.forceAdd(dID)
+}
+
+// SetAllocHealth is used to set the health of allocations for a deployment. If
+// there are any unhealthy allocations, the deployment is updated to be failed.
+// Otherwise the allocations are updated and an evaluation is created.
+func (w *Watcher) SetAllocHealth(req *structs.DeploymentAllocHealthRequest, resp *structs.DeploymentUpdateResponse) error {
+	watcher, err := w.getWatcher(req.DeploymentID)
+	if err != nil {
+		return err
+	}
+
+	return watcher.SetAllocHealth(req, resp)
+}
+
+// PromoteDeployment is used to promote a deployment. If promote is false,
+// deployment is marked as failed. Otherwise the deployment is updated and an
+// evaluation is created.
+func (w *Watcher) PromoteDeployment(req *structs.DeploymentPromoteRequest, resp *structs.DeploymentUpdateResponse) error {
+	watcher, err := w.getWatcher(req.DeploymentID)
+	if err != nil {
+		return err
+	}
+
+	return watcher.PromoteDeployment(req, resp)
+}
+
+// PauseDeployment is used to toggle the pause state on a deployment. If the
+// deployment is being unpaused, an evaluation is created.
+func (w *Watcher) PauseDeployment(req *structs.DeploymentPauseRequest, resp *structs.DeploymentUpdateResponse) error {
+	watcher, err := w.getWatcher(req.DeploymentID)
+	if err != nil {
+		return err
+	}
+
+	return watcher.PauseDeployment(req, resp)
+}
+
+// FailDeployment is used to fail the deployment.
+func (w *Watcher) FailDeployment(req *structs.DeploymentFailRequest, resp *structs.DeploymentUpdateResponse) error {
+	watcher, err := w.getWatcher(req.DeploymentID)
+	if err != nil {
+		return err
+	}
+
+	return watcher.FailDeployment(req, resp)
+}
+
+// createEvaluation commits the given evaluation to Raft but batches the commit
+// with other calls.
+func (w *Watcher) createEvaluation(eval *structs.Evaluation) (uint64, error) {
+	return w.evalBatcher.CreateEval(eval).Results()
+}
+
+// upsertJob commits the given job to Raft
+func (w *Watcher) upsertJob(job *structs.Job) (uint64, error) {
+	return w.raft.UpsertJob(job)
+}
+
+// upsertDeploymentStatusUpdate commits the given deployment update and optional
+// evaluation to Raft
+func (w *Watcher) upsertDeploymentStatusUpdate(
```

Comment:
What is the point of these methods? Seems like a pretty thin shim over the underlying layer?